proposals/reward-proposal-v2-reward-calculation.md

```diff
@@ -0,0 +1,101 @@
+is proposed. Let's call this approach `V1`. A prototype of the `V1` proposal is
+implemented. While the `V1` approach works functionally, it, however, has two
+main drawbacks.
+
+First, a restriction is enforced to prevent updates to all stake accounts and
+vote accounts during reward interval. In other words, the stake accounts and
+vote accounts are **locked** during reward interval. All stake account's actions
+such as delegate, redelegate, withdraw etc. are disallowed during reward
+interval and have to wait after reward interval.
+
+Second, there are no on-chain proof for rewards distribution at the epoch. In
+the `V1` approach, the rewards distribution is happening inside bank runtime,
+which update the accounts balance with epoch reward directly. Therefore, there
+is no on-chian records (i.e. transactions) for the rewards distribution. The
+rewards distributions is hidden and happens in a black-box manner.
+
+Given these two drawbacks, we are proposing an enhanced way to handle epoch
+reward - `V2`. The `V2` approach is based on `V1`. However, it completely
+decouples epoch reward into two distinct parts. The first part is to calculate
+the rewards. The second part is to distribute the rewards.
+
+In this proposal, we will focus on the first part - "calculate the reward". A
+follow-up proposal will talk about the second part - "distribute the reward".
+
+## Proposed Solutions
+
+Unlike the current epoch rewards computation approach, in which computing the
+rewards happens at the first slot of epoch boundary in the bank runtime
+directly, the new reward computation is moved to a background service in the
+proposed approach.
+
+This is very similar to the `V1` proposal, a separate service,
+"EpochRewardCalculationService" are created at the start time of the validator.
+This service is listening on a crossbeam channel for incoming rewards
+calculation requests, and handle the requests to compute the rewards.
+
+There are two main difference between the `V2` and `V1` approaches.
+
+The first difference  is that instead of sharing the `StakeCache` with the
+runtime during reward computation. The `StakeCache` is cloned at slot-height 0
+in the `V2` epoch. In this way, we can avoid locking stake accounts and vote
+accounts to prevent updates during the reward calculation.
+
+And the reward computation can last for the next `N` slot-hight. At the end of
+`Nth` slot-hight, the reward computation result should be available. In case
+that the reward result is not available yet at the `Nth` slot-height, the
+runtime will wait till the computation is finished before processing any blocks
+with height greater than `N`.
+
+The second difference is to introduce the the `reward proof`. To make sure that
+we have a consistent proof of the epoch rewards across different validator
+nodes, a merkle tree is computed from the epoch rewards of all stake and vote
```

Comment:
I'm not sure if a merkle tree is the right data structure for this, because of your last open question in the other proposal. The proofs may also bloat the redemption transactions, making it impossible to redeem many accounts at once. Why not re-run the inflation calculation on the account provided to see how much they should have earned? I imagine that would also be faster than a bunch of hashes. There's still the issue of what to do with deactivated stakes, since they can be reactivated before receiving the rewards :thinking:

Reply:
The idea of supplying the merkle proof in the redeem instructions is to make
sure that the reward to redeem does exists and is not corrupted in the
computed rewards. This will prevent that the redeemed rewards is different
from the computed rewards.  

If we re-run the calculation during reward, we could get a different result
from actual calculation and redeem a different reward. 

Yes. This can lead the issue of contention on the single reward proof account.
As you have suggested, we could split up "reward proof" account.

How about splitting it into one "proof" account and multiple "balance holder
accounts"? 

The "proof" account is readonly, holds the merkle root and total amount. The
actual balance transfer happens the individual balance account. 

The `DistributeReward` instruction will takes the following inputs.
    - accounts
        - target_account (W)
        - reward_balance_account (W)
        - reward_proof (R)
    - parameters
        - balance
        - merkle proof (a path of hashes)

This will help to relieve the contentions on the single reward proof account?